File: adapters/custom-engine/adapter.go
package main

import (
	"context"
	"fmt"
	"log"
	"sync"

	"github.com/usip/backend/schema"
)

const (
	adapterVersion = "custom-engine/0.1.0"
	engineID       = "custom-engine"
)

// runState tracks one in-flight or completed simulation run.
type runState struct {
	done    chan struct{}
	results *schema.SimResultsProto
	err     error
}

// adapterServer implements the SimAdapter gRPC service.
type adapterServer struct {
	schema.UnimplementedSimAdapterServer // forward-compat (no-op with require_unimplemented_servers=false)

	mu        sync.Mutex
	scenarios map[string]*schema.ScenarioProto
	runs      map[string]*runState
}

func newAdapterServer() *adapterServer {
	return &adapterServer{
		scenarios: make(map[string]*schema.ScenarioProto),
		runs:      make(map[string]*runState),
	}
}

func (a *adapterServer) Initialize(ctx context.Context, s *schema.ScenarioProto) (*schema.InitResponse, error) {
	if s.Id == "" {
		return &schema.InitResponse{Success: false, Message: "scenario id is required"}, nil
	}
	a.mu.Lock()
	a.scenarios[s.Id] = s
	a.mu.Unlock()

	log.Printf("Initialize: scenario=%s entities=%d missions=%d", s.Id, len(s.Entities), len(s.Missions))
	return &schema.InitResponse{
		Success:        true,
		Message:        fmt.Sprintf("scenario %q initialized", s.Name),
		AdapterVersion: adapterVersion,
	}, nil
}

func (a *adapterServer) Run(ctx context.Context, req *schema.RunRequest) (*schema.RunResponse, error) {
	a.mu.Lock()
	scenario, ok := a.scenarios[req.ScenarioId]
	if !ok {
		a.mu.Unlock()
		return &schema.RunResponse{Accepted: false, Message: "scenario not initialized; call Initialize first"}, nil
	}
	if _, exists := a.runs[req.RunId]; exists {
		a.mu.Unlock()
		return &schema.RunResponse{Accepted: false, Message: "run already exists"}, nil
	}
	state := &runState{done: make(chan struct{})}
	a.runs[req.RunId] = state
	a.mu.Unlock()

	// Execute asynchronously so Run returns promptly.
	go func() {
		defer close(state.done)
		log.Printf("Run: run=%s scenario=%s starting", req.RunId, req.ScenarioId)
		eng := newEngine(scenario, req.RunId)
		state.results = eng.run(req.ScenarioId, engineID, req.RunId)
		log.Printf("Run: run=%s done — %d tracks, %d events, %d kills",
			req.RunId, len(state.results.EntityTracks), len(state.results.Events), len(state.results.KillChains))
	}()

	return &schema.RunResponse{Accepted: true, Message: "run started"}, nil
}

func (a *adapterServer) GetResults(ctx context.Context, req *schema.ResultsRequest) (*schema.SimResultsProto, error) {
	a.mu.Lock()
	state, ok := a.runs[req.RunId]
	a.mu.Unlock()
	if !ok {
		return nil, fmt.Errorf("unknown run %q", req.RunId)
	}

	// Block until the run completes or the caller's context is cancelled.
	select {
	case <-state.done:
	case <-ctx.Done():
		return nil, ctx.Err()
	}
	if state.err != nil {
		return nil, state.err
	}
	return state.results, nil
}

func (a *adapterServer) Shutdown(ctx context.Context, req *schema.ShutdownRequest) (*schema.ShutdownResponse, error) {
	log.Printf("Shutdown requested (graceful=%v)", req.Graceful)
	return &schema.ShutdownResponse{Success: true}, nil
}

File: adapters/panopticon/adapter.py
"""SimAdapter gRPC service for the Panopticon adapter.

Mirrors the custom-engine adapter's contract: Initialize stores the scenario,
Run kicks off the sim asynchronously, GetResults blocks until it finishes.
"""

from __future__ import annotations

import logging
import threading

import grpc

from protostubs import adapter_pb2, adapter_pb2_grpc

log = logging.getLogger("panopticon.adapter")


class _RunState:
    def __init__(self):
        self.done = threading.Event()
        self.results = None
        self.error = None


class PanopticonAdapter(adapter_pb2_grpc.SimAdapterServicer):
    def __init__(self, engine, engine_id: str, version: str):
        self._engine = engine
        self._engine_id = engine_id
        self._version = version
        self._lock = threading.Lock()
        self._scenarios = {}
        self._runs = {}

    def Initialize(self, scenario, context):
        if not scenario.id:
            return adapter_pb2.InitResponse(success=False, message="scenario id is required")
        with self._lock:
            self._scenarios[scenario.id] = scenario
        log.info("Initialize: scenario=%s entities=%d missions=%d",
                 scenario.id, len(scenario.entities), len(scenario.missions))
        return adapter_pb2.InitResponse(
            success=True,
            message='scenario "%s" initialized' % scenario.name,
            adapter_version=self._version,
        )

    def Run(self, req, context):
        with self._lock:
            scenario = self._scenarios.get(req.scenario_id)
            if scenario is None:
                return adapter_pb2.RunResponse(
                    accepted=False, message="scenario not initialized; call Initialize first")
            if req.run_id in self._runs:
                return adapter_pb2.RunResponse(accepted=False, message="run already exists")
            state = _RunState()
            self._runs[req.run_id] = state

        log.info("Run: run=%s scenario=%s starting (engine=%s)",
                 req.run_id, req.scenario_id, self._engine.name)
        threading.Thread(
            target=self._execute, args=(scenario, req.run_id, state), daemon=True
        ).start()
        return adapter_pb2.RunResponse(accepted=True, message="run started")

    def _execute(self, scenario, run_id, state):
        try:
            state.results = self._engine.run(scenario, self._engine_id, run_id)
            log.info("Run: run=%s done — %d tracks, %d events, %d kills",
                     run_id, len(state.results.entity_tracks),
                     len(state.results.events), len(state.results.kill_chains))
        except Exception as exc:  # noqa: BLE001 - surfaced to the caller via GetResults
            log.exception("Run: run=%s failed", run_id)
            state.error = exc
        finally:
            state.done.set()

    def GetResults(self, req, context):
        with self._lock:
            state = self._runs.get(req.run_id)
        if state is None:
            context.abort(grpc.StatusCode.NOT_FOUND, "unknown run %s" % req.run_id)

        # Block until the run completes or the caller gives up.
        timeout = context.time_remaining()
        if not state.done.wait(timeout=timeout):
            context.abort(grpc.StatusCode.DEADLINE_EXCEEDED, "run still in progress")
        if state.error is not None:
            context.abort(grpc.StatusCode.INTERNAL, str(state.error))
        return state.results

    def Shutdown(self, req, context):
        log.info("Shutdown requested (graceful=%s)", req.graceful)
        return adapter_pb2.ShutdownResponse(success=True)
